README.md
# Snek
Another variant of the classic arcade game featuring an online leaderboard as well as offline functionality

Click [here](https://rahm.pythonanywhere.com/snek/) to try it out.


## Screenshots
<img src="https://user-images.githubusercontent.com/53840228/68999948-49a12380-08c8-11ea-906a-ac69863bad17.png" alt="Start Screen" width="200"/> <img src="https://user-images.githubusercontent.com/53840228/68999798-1eb5d000-08c6-11ea-9bf0-846b35ddc434.png" alt="Game Screen" width="200"/> <img src="https://user-images.githubusercontent.com/53840228/68999949-49a12380-08c8-11ea-97e3-df8b5a7d5935.png" alt="Leaderboard Screen" width="200"/>


## Getting Started
### Requirements:
- Python 3
- pip
- virtualenv 

### Installation
To get this project running on your local machine, the following steps need to be performed:

Clone this repository and navigate into the project:
```
git clone https://github.com/Rahmsauce/Snek.git
cd Snek
```

Create a virtual environment and activate it:
```
python3 -m venv env
source env/bin/activate
```

Install Flask:
```
pip install flask
```

Export and run application.py:
```
export FLASK_APP=application.py
export FLASK_ENV=development
flask run
```

Open http://localhost:5000 using a web browser.


changelog.txt
13.07.2020:
Getting Started
If there are no entries in the leaderboard, placeholder text is shown
Removed venv from repository
Script for removing all entries from the leaderboard

08.07.2020:
README updated

01.06.2020:
README updated


31.05.2020:
PWA conversion
Use of relative URLs in service_worker
Favicon


01.03.2020:
Small style and accessibility improvements on buttons and inputs


16.11.2019:
Added short delay between button clicks and action, so give button more time to finish their animations
Default 300ms click delay removed
README added


13.11.2019:
Version 2
Layout and Design overhaul
Translation to English
Removed settings page
Web app manifest
jQuery updated


29.12.2018:
Decreased sensibility of touch controls


18.12.2018:
Improved touch controls


16.12.2018:
Leaderboard table is created using javascript
Ajax-Request to enter leaderboard instead of form -> Page no longer needs to be refreshed


15.12.2018:
Leaderboard using Flask
New layout for starting page


05.10.2018:
Fixed focus bug in settings
Differentiation of primary and secondary buttons


30.07.2018:
V1


application.py
import os
from flask import Flask, render_template, request
import sqlite3
import json



app = Flask(__name__)

@app.route("/")
def index():
        return render_template("index.html")



@app.route("/service_worker.js")
def serviceWorker():
    return app.send_static_file("service_worker.js")



@app.route("/topten")
def topten():
        conn = sqlite3.connect(os.path.dirname(__file__) + "/leaderboard.db")
        cursor = conn.cursor()
        cursor.execute("SELECT name, score FROM leaderboard ORDER BY score DESC, rowid DESC")
        rows = cursor.fetchall()
        conn.close()

        for i in range(len(rows)):
                rows[i] = list(rows[i])

        return str(rows).replace("\'", "\"")



@app.route("/enterLeaderboard", methods=["POST"])
def enterLeaderboard():
        data = json.loads(str(request.get_data(as_text=True)))
        name = data["name"]
        score = data["score"]
        valid = True
        allowedCharacters = "abcdefghifklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"

        if len(name) == 0 or len(name) > 30:
                valid = False

        for letter in name:
                if letter not in allowedCharacters:
                        valid = False

        if int(score) != score:
                valid = False

        if valid:
                conn = sqlite3.connect(os.path.dirname(__file__) + "/leaderboard.db")
                cursor = conn.cursor()
                cursor.execute("INSERT INTO leaderboard (name, score) VALUES (?,?)", (name, score))
                cursor.execute("DELETE FROM leaderboard WHERE rowid NOT IN (SELECT rowid FROM leaderboard ORDER BY score DESC, rowid DESC LIMIT 10)")
                conn.commit()
                conn.close()

        return ("", 204)


static/service_worker.js
self.addEventListener("install", (event) => {
    event.waitUntil(
        caches.open("v1").then((cache) => {
            return cache.addAll([
                ".",
                "static/styles.css",
                "static/fontello/css/fontello.css",
                "static/jquery-3.4.1.min.js",
                "static/script.js",
                "static/PressStart2P.ttf",
                "static/fontello/font/fontello.woff2?84098909",
                "static/manifest.json",
                "static/icon_512.png"
            ]);
        })
    );
});



self.addEventListener("fetch", (event) => {
    if (event.request.url.indexOf("topten") == -1 && event.request.url.indexOf("enterLeaderboard") == -1) {
        event.respondWith(
            caches.match(event.request)
        );
    }

    else {
        event.respondWith(
            fetch(event.request)
        );
    }
});


static/script.js
var grid;
var gridSize = 10;
var speed = 150;
var snakeLength;
var movingDirection = "right";
var moveInterval;
var score;
var topten;

var touchstartPosition;



window.onload = function() {
    registerServiceWorker();
	refreshLeaderboardScreen();
	addDirectionChangeEvents();

	$("#settingsScreen input, #enterLeaderboardScreen input").on("click", function(e) {
		e.target.focus();
	});



    function registerServiceWorker() {
        if ("serviceWorker" in navigator) {
        	navigator.serviceWorker.register("service_worker.js");
        }
    }


	function addDirectionChangeEvents() {
		$(document).on("keydown", function(e) {
			if (e.which == 37) {
				movingDirection = "left";
			}
			else if (e.which == 38) {
				movingDirection = "up";
			}
			else if (e.which == 39) {
				movingDirection = "right";
			}
			else if (e.which == 40) {
				movingDirection = "down";
			}
		});



		window.ontouchstart = function(e) {
			touchstartPosition = {
				x: e.touches[0].clientX,
				y: e.touches[0].clientY
			}
		};

		window.ontouchmove = function(e) {
			var touchmovePosition = {
				x: e.changedTouches[0].clientX,
				y: e.changedTouches[0].clientY
			};


			var xDifference = touchstartPosition.x - touchmovePosition.x;
			var yDifference = touchstartPosition.y - touchmovePosition.y;
			var direction = "none";

			if (Math.abs(xDifference) > 50) {
				if (xDifference < 0) {
					movingDirection = "right";
				}
				else {
					movingDirection = "left";
				}
			}
			else if (Math.abs(yDifference) > 50) {
				if (yDifference < 0) {
				movingDirection = "down";
				}
				else {
					movingDirection = "up";
				}
			}
		};
	}
}





var starter = {
	prepare: function() {
		score = 0;

		this.createGrid(gridSize);
		this.createSnake();
		createFood();
		display();

		$(document).on("keydown touchstart", starter.start);
		setTimeout(function() {
			$(".popup").fadeOut(200);
		}, 200);
	},

	createGrid: function(size) {
		grid = [];
		$("#grid > div").remove();

		for (var i=0;i<size*size;i++) {
			grid.push({type:"empty"});
		}

		for (field in grid) {
			var newField = $("<div></div>");

			$("#grid").append(newField);
		}

		$("#grid").css({
			"grid-template-rows": "repeat(" + gridSize + ", 1fr)",
			"grid-template-columns": "repeat(" + gridSize + ", 1fr)"
		});

	},
	createSnake: function() {
		snakeLength = 1;
		grid[gridSize + 1] = {type: "snake", disappearIn: 1, head: true};
	},

	start: function() {
		moveInterval = setInterval(function() {
			if (move(movingDirection)) {
				display();
			}
			else {
				$("#gameOverScoreLabel").text("Score: " + score);
				setTimeout(function() {
					$("#startScreen").fadeIn(200);
				}, 200);
				$("#startScreen h1").text("Game Over");
				$("#startScreen #gameOverScoreLabel").css("display", "block");
				clearInterval(moveInterval);

				if (topten.length < 10) {
					setTimeout(function() {
						$("#enterLeaderboardScreen").fadeIn(200);
					}, 200);
					$("#scoreInput").val(score);
				}
				else {
					for (var i=topten.length-1;i>-1;i--) {
						if (score >= topten[i][1]) {
							setTimeout(function() {
								$("#enterLeaderboardScreen").fadeIn(200);
							}, 200);

							$("#scoreInput").val(score);
							break;
						}
					}
				}
			}
		}, speed);

		$(document).unbind("keydown", starter.start);
		$(document).unbind("touchstart", starter.start);
	}
};




function createFood() {
	var foodCreated = false;

	while (foodCreated == false && score+1 != gridSize*gridSize) {
		var randomNumber = Math.floor(Math.random() * (grid.length));

		if (grid[randomNumber].type == "empty") {
			grid[randomNumber].type = "food";
			foodCreated = true;
		}
	}
}
function growSnake() {
	for (field in grid) {
		if (grid[field].type == "snake") {
			grid[field].disappearIn += 1;
		}
	}

	snakeLength++;
	score++;
	createFood();
}





function move(direction) {
	var currentHeadPosition;
	var nextFieldIndex = false;
	var stillAlive = false;


	for (var i=0;i<grid.length;i++) {
		if(grid[i].type == "snake") {
			if (grid[i].head === true) {
				currentHeadPosition = i;
				grid[i].head = undefined;
			}


			grid[i].disappearIn -= 1;

			if (grid[i].disappearIn == 0) {
				grid[i].type = "empty";
				grid[i].disappearIn = undefined;
			}
		}
	}


	if (direction == "up") {
		if (currentHeadPosition >= gridSize) {
			nextFieldIndex = currentHeadPosition - gridSize;
		}
	}
	else if (direction == "right") {
		if (currentHeadPosition % gridSize != gridSize -1) {
			nextFieldIndex = currentHeadPosition + 1;
		}
	}
	else if (direction == "down") {
		if (currentHeadPosition < gridSize * gridSize - gridSize) {
			nextFieldIndex = currentHeadPosition + gridSize;
		}
	}
	else if (direction == "left") {
		if (currentHeadPosition % gridSize != 0) {
			nextFieldIndex = currentHeadPosition - 1;
		}
	}


	if (nextFieldIndex !== false) {
		if (grid[nextFieldIndex].type != "snake") {
			if (grid[nextFieldIndex].type == "food") {
				growSnake();
			}

			grid[nextFieldIndex].type = "snake";
			grid[nextFieldIndex].head = true;
			grid[nextFieldIndex].disappearIn = snakeLength;
			stillAlive = true;
		}
	}

	return stillAlive;
}





function display() {
	$("#grid > div").removeClass();

	for (var i=0;i<grid.length;i++) {
		if (grid[i].type == "snake") {
			$("#grid > div").eq(i).addClass("snake");

			if (grid[i].head == true) {
				$("#grid > div").eq(i).addClass("head");
			}
		}
		if (grid[i].type == "food") {
			$("#grid > div").eq(i).addClass("food");
		}
	}


	$("#scoreLabel").text(score);
}



function enterLeaderboard() {
	var data = {
		name: $("#nameInput").val(),
		score: score
	}


	valid = validate(data.name, data.score);

	if (valid === true) {
		var xhttp = new XMLHttpRequest();
		xhttp.open("POST", window.location.href + "enterLeaderboard");
		xhttp.setRequestHeader("Content-type", "application/x-www-form-urlencoded");
		xhttp.send(JSON.stringify(data));

		xhttp.onreadystatechange = function() {
			if (this.readyState == 4 && this.status >= 200 && this.status < 300) {
				closeLeaderboardScreen();
			}

			else if (this.readyState == 4 && this.status == 0) {
				$("#statusIndicator").text("Error: No connection").css("display", "block");
			}
		};
	}
	else {
		$("#statusIndicator").text(valid).css("display", "block");
	}


	function validate(name, score) {
		var valid = true;
		var allowedCharacters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";

		if (name.length > 30 || name.length == 0) {
			valid = "The name must be between 1 and 30 characters long.";
		}

		for (letter in name) {
			if (allowedCharacters.indexOf(name[letter]) == -1) {
				valid = "The name must not contain any special characters.";
			}
		}

		if (parseInt(score) != score) {
			valid = "Invalid score";
		}

		return valid
	}
}

function closeLeaderboardScreen() {
	setTimeout(function() {
		$("#enterLeaderboardScreen").fadeOut(200);
	}, 200);

	setTimeout(function() {
		$("#statusIndicator").text("").css("display", "none");
	}, 400);
}



function refreshLeaderboardScreen() {
	var xhttp = new XMLHttpRequest();

	xhttp.onreadystatechange = function() {
		if (this.readyState == 4 && this.status == 200) {
			topten = JSON.parse(xhttp.responseText);

			$("#leaderboardScreen table").html("");

            if (topten.length == 0) {
                $("#leaderboardScreen table").html("No entries yet");
            }

            else {
    			for (var i=0;i<topten.length;i++) {
	    			var row = $("<tr></tr>");
	    			var position = $("<td>" + (i+1) + "</td>");
	    			var name = $("<td>" + topten[i][0] + "</td>");
	    			var score = $("<td>" + topten[i][1] + "</td>");

	    			$(row).append(position, name, score);
	    			$("#leaderboardScreen table").append(row);
	    		}
            }
		}

		else if (this.status == 0) {
			$("#leaderboardScreen table").html("<p style='color: #d66;'>Error: No connection</p>");
		}
	};

	xhttp.open("GET", window.location.href + "topten");
	xhttp.send();
}
